src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs

```diff
@@ -21,66 +21,53 @@ namespace Microsoft.CodeAnalysis.Editor.Shared.Tagging;
-        // Avoid reallocating in the case where we're on the same snapshot.
-        => originalTagSpan.Span.Snapshot == textSnapshot
-            ? originalTagSpan
-            : GetTranslatedTagSpan(originalTagSpan, textSnapshot, _spanTrackingMode);
+    private static SnapshotSpan GetTranslatedSpan(TagSpan<TTag> originalTagSpan, ITextSnapshot textSnapshot, SpanTrackingMode trackingMode)
+        // SnapshotSpan no-ops if you pass it the same snapshot that it is holding onto.
+        => originalTagSpan.Span.TranslateTo(textSnapshot, trackingMode);
+
+    private TagSpan<TTag> GetTranslatedTagSpan(TagSpan<TTag> originalTagSpan, ITextSnapshot textSnapshot)
+        => GetTranslatedTagSpan(originalTagSpan, textSnapshot, _spanTrackingMode);
 
     private static TagSpan<TTag> GetTranslatedTagSpan(TagSpan<TTag> originalTagSpan, ITextSnapshot textSnapshot, SpanTrackingMode trackingMode)
         // Avoid reallocating in the case where we're on the same snapshot.
-        => originalTagSpan is TagSpan<TTag> tagSpan && tagSpan.Span.Snapshot == textSnapshot
-            ? tagSpan
+        => originalTagSpan.Span.Snapshot == textSnapshot
+            ? originalTagSpan
             : new(GetTranslatedSpan(originalTagSpan, textSnapshot, trackingMode), originalTagSpan.Tag);
 
-    public ITextBuffer Buffer => _textBuffer;
-
-    public SpanTrackingMode SpanTrackingMode => _spanTrackingMode;
-
     public bool HasSpanThatContains(SnapshotPoint point)
-    {
-        var snapshot = point.Snapshot;
-        Debug.Assert(snapshot.TextBuffer == _textBuffer);
-
-        return _tree.HasIntervalThatContains(point.Position, length: 0, new IntervalIntrospector(snapshot, _spanTrackingMode));
-    }
+        => _tree.HasIntervalThatContains(point.Position, length: 0, new IntervalIntrospector(point.Snapshot, _spanTrackingMode));
 
-    public IReadOnlyList<TagSpan<TTag>> GetIntersectingSpans(SnapshotSpan snapshotSpan)
-        => SegmentedListPool<TagSpan<TTag>>.ComputeList(
-            static (args, tags) => args.@this.AppendIntersectingSpansInSortedOrder(args.snapshotSpan, tags),
-            (@this: this, snapshotSpan));
+    public bool HasSpanThatIntersects(SnapshotPoint point)
+        => _tree.HasIntervalThatIntersectsWith(point.Position, new IntervalIntrospector(point.Snapshot, _spanTrackingMode));
 
     /// <summary>
     /// Gets all the spans that intersect with <paramref name="snapshotSpan"/> in sorted order and adds them to
     /// <paramref name="result"/>.  Note the sorted chunk of items are appended to <paramref name="result"/>.  This
     /// means that <paramref name="result"/> may not be sorted if there were already items in them.
     /// </summary>
-    private void AppendIntersectingSpansInSortedOrder(SnapshotSpan snapshotSpan, SegmentedList<TagSpan<TTag>> result)
+    public void AddIntersectingTagSpans(SnapshotSpan snapshotSpan, SegmentedList<TagSpan<TTag>> result)
```

Comment:
replacement for GetSpans.  No allocs.  Caller passes in the value the ywant to fill.